File: src/apis/diet-planner/DietPlanner.Api/Domain/Product.cs
namespace DietPlanner.Api.Domain;

public class Product
{
    public Guid Id { get; set; }
    public required string Name { get; set; }
    public decimal? CaloriesPer100g { get; set; }
    public decimal? ProteinPer100g { get; set; }
    public decimal? CarbsPer100g { get; set; }
    public decimal? FatPer100g { get; set; }
    public decimal? FiberPer100g { get; set; }
    public string DefaultUnit { get; set; } = "g";
    public decimal? DensityGramsPerMl { get; set; }  // For accurate volume conversions
    public decimal? GramPerPiece { get; set; }       // For "piece" units
    public required string CreatedByUserId { get; set; }
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime? UpdatedAt { get; set; }
    public DateTime? DeletedAt { get; set; }

    // Navigation properties
    public ICollection<RecipeIngredient> RecipeIngredients { get; set; } = new List<RecipeIngredient>();

    // Computed property
    public bool IsDeleted => DeletedAt.HasValue;
}

File: src/ui/src/test/factories/product.ts
export interface ProductFixture {
  id: string;
  name: string;
  caloriesPer100g: number;
  proteinPer100g: number;
  carbsPer100g: number;
  fatPer100g: number;
  fiberPer100g: number | null;
  defaultUnit: string;
  densityGramsPerMl: number | null;
  gramPerPiece: number | null;
  isOwner: boolean;
  createdAt: string;
}

let counter = 0;

export function productFactory(overrides: Partial<ProductFixture> = {}): ProductFixture {
  counter += 1;
  return {
    id: `product-${String(counter)}-0000-0000-0000-000000000000`,
    name: `Product ${String(counter)}`,
    caloriesPer100g: 100,
    proteinPer100g: 10,
    carbsPer100g: 15,
    fatPer100g: 5,
    fiberPer100g: null,
    defaultUnit: 'g',
    densityGramsPerMl: null,
    gramPerPiece: null,
    isOwner: true,
    createdAt: new Date().toISOString(),
    ...overrides,
  };
}
